File: src/main/kotlin/com/example/features/ControlFlowDemo.kt
package com.example.features

/**
 * Demonstrates control-flow constructs:
 *   - `for` over a numeric range (`..`) with literal and variable bounds
 *   - `for` with `until`, `downTo`, and `step`
 *   - `while` loop
 *   - `if` expression used as a value (inline ternary)
 *   - String template concatenation
 *
 * `demoControlFlow()`, `demoRanges()` and `demoStringTemplate()` are global
 * functions accessible from Main.lua after `require("features/ControlFlowDemo")`.
 */
class ControlFlowDemo

fun demoControlFlow(): Int {
    var sum = 0
    for (i in 1..5) {
        sum += i
    }
    var x = 10
    while (x > 5) {
        x = x - 1
    }
    // if expression: (sum == 15 and sum + x or 0) → 15 + 5 = 20
    val result = if (sum == 15) { sum + x } else 0
    return result
}

/** Demonstrates `until`, `downTo`, `step`, and variable-bound ranges. */
fun demoRanges(): Int {
    var total = 0
    // until: 0, 1, 2, 3, 4  → sum = 10
    for (i in 0 until 5) {
        total += i
    }
    // downTo: 5, 4, 3, 2, 1  → sum += 15  → total = 25
    for (i in 5 downTo 1) {
        total += i
    }
    // step: 0, 2, 4  → sum += 6  → total = 31
    for (i in 0..4 step 2) {
        total += i
    }
    // variable bounds
    val lo = 1
    val hi = 3
    for (i in lo..hi) {
        total += i  // 1+2+3 = 6  → total = 37
    }
    return total
}

fun demoStringTemplate(): String {
    val lang = "Lua"
    val version = 5
    return "Running on $lang ${version + 0}"
}

File: outputDir/vscripts/features/ControlFlowDemo.lua
-- package: com.example.features

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "features/ControlFlowDemo.kt", {["1-15"]=1,["16"]=17,["17"]=18,["18-19"]=19,["20"]=21,["21"]=22,["22-23"]=23,["24"]=26,["25-28"]=27,["29"]=32,["30"]=34,["31-32"]=35,["33"]=38,["34-35"]=39,["36"]=42,["37-38"]=43,["39"]=46,["40"]=47,["41"]=48,["42-43"]=49,["44-47"]=51,["48"]=55,["49"]=56,["50-52"]=57})

ControlFlowDemo = {}
ControlFlowDemo.__index = ControlFlowDemo

function ControlFlowDemo:new()
    local self = setmetatable({}, ControlFlowDemo)
    return self
end


function demoControlFlow()
    local sum = 0
    for i = 1, 5 do
        sum = ktox_plusAssign(sum, i)
    end
    local x = 10
    while x > 5 do
        x = x - 1
    end
    local result = (sum == 15 and sum + x or 0)
    return result
end

function demoRanges()
    local total = 0
    for i = 0, 5 - 1 do
        total = ktox_plusAssign(total, i)
    end
    for i = 5, 1, -1 do
        total = ktox_plusAssign(total, i)
    end
    for i = 0, 4, 2 do
        total = ktox_plusAssign(total, i)
    end
    local lo = 1
    local hi = 3
    for i = lo, hi do
        total = ktox_plusAssign(total, i)
    end
    return total
end

function demoStringTemplate()
    local lang = "Lua"
    local version = 5
    return "Running on " .. tostring(lang) .. " " .. tostring(version + 0)
end

